// typegoose/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose = require("mongoose");
const data_1 = require("./data");
function isPrimitive(Type) {
    return ['String', 'Number', 'Boolean', 'Date', 'Decimal128', 'ObjectID'].includes(Type.name);
}
exports.isPrimitive = isPrimitive;
function isObject(Type) {
    let prototype = Type.prototype;
    let name = Type.name;
    while (name) {
        if (name === 'Object') {
            return true;
        }
        prototype = Object.getPrototypeOf(prototype);
        name = prototype ? prototype.constructor.name : null;
    }
    return false;
}
exports.isObject = isObject;
function isNumber(Type) {
    return Type.name === 'Number';
}
exports.isNumber = isNumber;
function isString(Type) {
    return Type.name === 'String';
}
exports.isString = isString;
function initAsObject(name, key) {
    if (!data_1.schema[name]) {
        data_1.schema[name] = {};
    }
    if (!data_1.schema[name][key]) {
        data_1.schema[name][key] = {};
    }
}
exports.initAsObject = initAsObject;
function initAsArray(name, key) {
    if (!data_1.schema[name]) {
        data_1.schema[name] = {};
    }
    if (!data_1.schema[name][key]) {
        data_1.schema[name][key] = [{}];
    }
}
exports.initAsArray = initAsArray;
function getClassForDocument(document) {
    const modelName = document.constructor.modelName;
    return data_1.constructors[modelName];
}
exports.getClassForDocument = getClassForDocument;
function isDocument(doc) {
    return doc instanceof mongoose.Model;
}
exports.isDocument = isDocument;
function isDocumentArray(docs) {
    return Array.isArray(docs) && docs.every(v => isDocument(v));
}
exports.isDocumentArray = isDocumentArray;
